kafka/main.go

```diff
@@ -67,94 +87,154 @@ func StartConsumers(providedServer *server.Server, logger *zerolog.Logger) error
 		topics = append(topics, topicMapping.Topic)
 	}
 
-	consumerCount, err := strconv.Atoi(os.Getenv("KAFKA_CONSUMERS_PER_NODE"))
+	reader := newConsumer(topics, adsConsumerGroupV1, logger)
+
+	batchPipeline := make(chan *MessageContext)
+	ctx := context.Background()
+	go processMessagesIntoBatchPipeline(ctx, topicMappings, providedServer, reader, batchPipeline, logger)
+	for {
+		err := readAndCommitBatchPipelineResults(ctx, reader, batchPipeline, logger)
+		if err != nil {
+			// If readAndCommitBatchPipelineResults returns an error.
+			close(batchPipeline)
+			return err
+		}
+	}
+}
+
+// readAndCommitBatchPipelineResults does a blocking read of the batchPipeline channel and
+// then does a blocking read of the errorResult in the MessageContext in the batchPipeline.
+// When an error appears it means that the channel was closed or a temporary error was
+// encountered. In the case of a temporary error, the application returns an error without
+// committing so that the next reader gets the same message to try again.
+func readAndCommitBatchPipelineResults(
```

Comment:
When this func errors how do we restart the consumer? If we have a bad message will this block all processing until is fixed?

Reply:
could this lead to us reading the same message again and failing again. Just wondering if we could just log or dlq the message and commit.

Reply:
Reprocessing the same message again is the intended behavior. This is why it is critical to categorize temporary and permanent failures correctly. Permanent errors need to be logged and proceeded from. Temporary errors must be retried until they either fail permanently or succeed.